dds_web/security/project_user_keys.py

```diff
@@ -1,17 +1,128 @@
+
+
+def generate_project_key_pair(user):
+    private_key = asymmetric.x25519.X25519PrivateKey.generate()
+    private_key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PrivateFormat.Raw,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+    encrypted_private_key = encrypt_project_private_key_for_user(user, private_key_bytes)
+    del private_key_bytes
+    del private_key
+    gc.collect()
+    return {
+        "public_key": private_key.public_key()
+        .public_bytes(encoding=serialization.Encoding.Raw, format=serialization.PublicFormat.Raw)
+        .hex()
+        .upper(),
+        "encrypted_private_key": encrypted_private_key,
+    }
+
+
+def encrypt_with_aes(key, plaintext, aad=None):
+    """
+    Encrypts the plaintext with the AES algorithm using the key
+    :param key: symmetric key of the user
+    :param plaintext: a byte string
+    :param aad: Additional data that should be authenticated with the key, but is not encrypted. Can be None.
+    """
+    aesgcm = ciphers.aead.AESGCM(key)
+    nonce = os.urandom(12)
+    return {"nonce": nonce, "encrypted_key": aesgcm.encrypt(nonce, plaintext, aad)}
+
+
+def decrypt_with_aes(key, ciphertext, nonce, aad=None):
+    """
+    Decrypts the ciphertext with the AES algorithm using the key
+    :param key: symmetric key of the user
+    :param ciphertext: encrypted content
+    :param nonce: a unique value that has to be used only once to encrypt a given data with the given key
+    :param aad: Additional data that should be authenticated with the key, but is not encrypted. Can be None.
+    """
+    aesgcm = ciphers.aead.AESGCM(key)
+    return aesgcm.decrypt(nonce, ciphertext, aad)
+
+
+def encrypt_user_private_key(user, private_key):
+    user.temporary_key = ciphers.aead.AESGCM.generate_key(bit_length=256)
+    encrypted_key_dict = encrypt_with_aes(
+        user.temporary_key, private_key, aad=b"private key for user " + user.username.encode()
+    )
+    user.nonce = encrypted_key_dict["nonce"]
+    return encrypted_key_dict["encrypted_key"]
+
+
+def decrypt_user_private_key(user):
+    if user.temporary_key and user.private_key and user.nonce:
+        return decrypt_with_aes(
+            user.temporary_key,
+            user.private_key,
+            user.nonce,
+            aad=b"private key for user " + user.username.encode(),
+        )
+    exception = Exception("User account is not properly setup!")
+    flask.current_app.logger.exception(exception)
+    raise exception
+
+
+def generate_user_key_pair(user):
     private_key = asymmetric.rsa.generate_private_key(public_exponent=65537, key_size=4096)
+    private_key_bytes = private_key.private_bytes(
+        serialization.Encoding.DER, serialization.PrivateFormat.PKCS8, serialization.NoEncryption()
+    )
+    encrypted_private_key = encrypt_user_private_key(user, private_key_bytes)
```

Comment:
You need to call the `private_key.public_key()` here (atm `del private_key` is before using it in the return), that's one of the reasons the tests fail atm.